### PR TITLE
Validate that all context types match

### LINKED ISSUE #21: Allow Grats to validate context values
The third argument to a resolver is always the request context. This can be anything the user configures, so it's generally hard to typecheck effectively. However, Grats knows each function/method that is a GraphQL field (thanks to the `@gqlField` annotations) so could be charged with ensuring all fields are typed in a way that matches the developer's expectations.

This would require some way for Grats users to inform users what type they are passing to their executor. I'll need to brainstorm what options make sense here:

1. Grats config file allows you to define a module path + export name
2. Some well defined global type (`GratsContextValue`?)  that can be defined by the user in their code and Grats will look for
3. Some other TypeScript trick that I'm not thinking of?
4. _Edit_ `/** @gqlContext */` on the context type declaration.

We should also examine how existing TS GraphQL servers address this. I'm thinking specifically of TypeGraphQL and Pothos.

### PR DESCRIPTION
**Pull request by captbaritone**
Fixes #21. Directly implements what @mandx suggested in https://github.com/captbaritone/grats/issues/21#issuecomment-1789986205.

An alternative `/** @gqlContext */` approach is explored in #24.

### CODE REVIEW

File: TODO.md [hunk 1]
```diff
@@ -50,7 +50,6 @@
 - [ ] Improve playground
   - Could we actually evaluate the resolvers? Maybe in a worker?
   - Could we hook up GraphiQL 2?
-- [ ] Can we ensure the context and ast arguments of resolvers are correct?
```
Comment (captbaritone) [inline TODO.md:53]: Maybe premature to add this since we don't yet typecheck the `info` argument.

File: src/Extractor.ts [hunk 2]
```diff
@@ -980,7 +985,7 @@ export class Extractor {
     } else if (node.kind === ts.SyntaxKind.FalseKeyword) {
       return this.gql.boolean(node, false);
     } else if (ts.isObjectLiteralExpression(node)) {
-      return this.cellectObjectLiteral(node);
+      return this.collectObjectLiteral(node);
```
Comment (captbaritone) [inline src/Extractor.ts:988]: Unrelated typo fix.

File: src/tests/fixtures/resolver_context/ContextValueTypedAsString.invalid.ts.expected [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    return ctx;
+  }
+}
+
+-----------------
+OUTPUT
+-----------------
+src/tests/fixtures/resolver_context/ContextValueTypedAsString.invalid.ts:4:30 - error: Expected context parameter's type to be a type reference Grats validates that your context parameter is type-safe by checking all context values reference the same type declaration.
+
+4   greeting(args: never, ctx: string): string {
```
Comment (mandx) [inline src/tests/fixtures/resolver_context/ContextValueTypedAsString.invalid.ts.expected:17]: Is not supporting `string` intentional? Or is it because some technical limitation?

File: src/tests/fixtures/resolver_context/ContextValueTypedAsNever.ts.expected [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    return ctx.greeting;
+  }
+}
+
+-----------------
+OUTPUT
+-----------------
+src/tests/fixtures/resolver_context/ContextValueTypedAsNever.ts:4:30 - error: Expected context parameter's type to be a type reference Grats validates that your context parameter is type-safe by checking all context values reference the same type declaration.
+
+4   greeting(args: never, ctx: never): string {
```
Comment (mandx) [inline src/tests/fixtures/resolver_context/ContextValueTypedAsNever.ts.expected:17]: Is it not possible to allow `unknown` as the type for `ctx`?

**Review by mandx — COMMENTED** [review]
Very exciting this change! I was thinking (maybe you though this too) Grats could also support the `@gqlContext` docblock tag, so when present, all `ctx` arguments' type must resolve to that, otherwise, use this checking strategy as fallback.

File: src/tests/fixtures/resolver_context/ContextValueTypedAsString.invalid.ts.expected (hunk 3 above)
Comment (captbaritone) [reply src/tests/fixtures/resolver_context/ContextValueTypedAsString.invalid.ts.expected:17]: Technical limitation because we are tracking the _definition_ of the type, and primitives and literals won't work.

File: src/tests/fixtures/resolver_context/ContextValueTypedAsNever.ts.expected (hunk 4 above)
Comment (captbaritone) [reply src/tests/fixtures/resolver_context/ContextValueTypedAsNever.ts.expected:17]: We could, just haven't added it yet! Unclear if it's needed though.  I guess if you need to access the `info` argument after the context value and you don't actually ever use the context value?

**Comment by captbaritone** [comment]
> Very exciting this change! I was thinking (maybe you though this too) Grats could also support the @gqlContext docblock tag, so when present, all ctx arguments' type must resolve to that, otherwise, use this checking strategy as fallback.

Yeah, I had a similar thought but more sequential than options. We can start here, and if we end up needing to actually know which value is the context value (say for positional arguments) we could layer on the `@gqlContext` tag as a new requirement. It would be a breaking change, but it would just mean adding a single tag to your single context type definition.

**Comment by mandx** [comment]
> > Very exciting this change! I was thinking (maybe you though this too) Grats could also support the @gqlContext docblock tag, so when present, all ctx arguments' type must resolve to that, otherwise, use this checking strategy as fallback.
> 
> Yeah, I had a similar thought but more sequential than options. We can start here, and if we end up needing to actually know which value is the context value (say for positional arguments) we could layer on the `@gqlContext` tag as a new requirement. It would be a breaking change, but it would just mean adding a single tag to your single context type definition.

I think all modes are valid, it just adds some complexity to the checking code:
* If a positional arguments are found, then `@gqlContext` is required to be defined, and all present `context` args must reference that type.
* If no positional args are found, but `@gqlContext` is defined, then all present `context` args must reference that type.
* If no positional args are found, and `@gqlContext` is not defined, then just check that all present `context` args must reference that the same type, whatever that is.

I can see this working great, as long as if a check fails the error message is descriptive enough to explain what and why, and provide actions or suggestions.

File: src/tests/fixtures/resolver_context/ContextValueTypedAsNever.ts.expected (hunk 4 above)
Comment (mandx) [reply src/tests/fixtures/resolver_context/ContextValueTypedAsNever.ts.expected:17]: Yeah, that's one use case. It would be a just escape hatch, in case we need to add the context argument (for whatever reason) but we don't want to deal with typing the context _right_ now.

However, if it's easy to add, we could not add it now, and wait for (less theoretical, more practical) reasons to add it and make a better decision then.

File: src/Errors.ts [hunk 5]
```diff
@@ -161,11 +161,11 @@ export function typeNameDoesNotMatchExpected(expected: string) {
 }
 
 export function argumentParamIsMissingType() {
-  return "Expected GraphQL field arguments to have a TypeScript type. If there are no arguments, you can use `args: never`.";
+  return "Expected GraphQL field arguments to have a TypeScript type. If there are no arguments, you can use `args: unknown`.";
```
Comment (captbaritone) [inline src/Errors.ts:164]: This check/guidance was wrong before, but I'm fixing it here. Should have been `unknown` for this use case not `never`.

File: website/docusaurus.config.js [hunk 6]
```diff
@@ -22,7 +22,7 @@ const config = {
   projectName: "grats", // Usually your repo name.
 
   onBrokenLinks: "throw",
-  onBrokenMarkdownLinks: "warn",
+  onBrokenMarkdownLinks: "throw",
```
Comment (captbaritone) [inline website/docusaurus.config.js:25]: Errors were turning up in CI but not local dev. This fixes that.
